src/bin/pg_autoctl/keeper.c

```diff
@@ -577,6 +546,65 @@ keeper_update_pg_state(Keeper *keeper)
 }
 
 
+/*
+ * keeper_state_check_postgres checks that the Postgres control data main
+ * properties are still as we expect them to be. At the moment we don't support
+ * Postgres minor and major upgrades, and we can't support the system
+ * identifier ever changing.
+ */
+static bool
+keeper_state_check_postgres(Keeper *keeper, PostgresControlData *control)
+{
+	KeeperStateData *keeperState = &(keeper->state);
+
+	/*
+	 * We got new control data from either running pg_controldata or connecting
+	 * to the local Postgres instance and running our
+	 * pgsql_get_postgres_metadata() SQL query. In either case we now need to
+	 * update our Keeper State with the control data values.
+	 */
+	if (keeperState->system_identifier != control->system_identifier &&
+		keeperState->system_identifier != 0)
+	{
+		/*
+		 * This is a physical replication deal breaker, so it's mighty
+		 * confusing to get that here. In the least, the keeper should get
+		 * initialized from scratch again, but basically, we don't know what we
+		 * are doing anymore.
+		 */
+		log_error("Unknown PostgreSQL system identifier: %" PRIu64 ", "
+																   "expected %" PRIu64,
+				  keeperState->system_identifier,
+				  control->system_identifier);
+		return false;
+	}
+
+	if (keeperState->pg_control_version != control->pg_control_version &&
+		keeperState->pg_control_version != 0)
+	{
+		/* Postgres minor upgrade happened */
+		log_warn("PostgreSQL version changed from %u to %u",
+				 keeperState->pg_control_version,
+				 control->pg_control_version);
+	}
+
+	if (keeperState->catalog_version_no != control->catalog_version_no &&
+		keeperState->catalog_version_no != 0)
+	{
+		/* Postgres major upgrade happened */
+		log_warn("PostgreSQL catalog version changed from %u to %u",
+				 keeperState->catalog_version_no,
+				 control->catalog_version_no);
+	}
+
+	keeperState->pg_control_version = control->pg_control_version;
```

Comment:
I'd still move these 3  lines out of the function like before as  the name of the function is `keeper_state_check_postgres `  ( e.g., setting these values seem like an unexpected side effect from a function name `*check*`)

Or (less preferable), rename it to something like `keeper_state_check_and_set_postgres`

Reply:
Good point, done now.